cpu/samd5x/periph/can.c

```diff
@@ -0,0 +1,768 @@
+        _exit_init_mode(dev->conf->can);
+    }
+
+    return idx;
+}
+
+static int _remove_filter(candev_t *candev, const struct can_filter *filter)
+{
+    can_t *dev = container_of(candev, can_t, candev);
+
+    int16_t idx = 0;
+    bool _filter_exists = false;
+    if (filter->can_id & CAN_EFF_FLAG) {
+        _filter_exists = _find_filter(dev, filter, false, &idx);
+        if (_filter_exists) {
+            DEBUG("Extended filter to disable at idx = %d\n", idx);
+            dev->msg_ram_conf.ext_filter[idx].XIDFE_0.bit.EFEC = CANDEV_SAMD5X_FILTER_DISABLE;
+
+            for (uint8_t i = 0; i < ARRAY_SIZE(dev->msg_ram_conf.ext_filter); i++) {
+                DEBUG("can->msg_ram_conf.std_filter[%u] = 0x%08lx, filter conf = %u\n", i,
+                        (uint32_t)(dev->msg_ram_conf.ext_filter[i].XIDFE_0.bit.EFID1), dev->msg_ram_conf.ext_filter[i].XIDFE_0.bit.EFEC);
+            }
+        }
+        else {
+            DEBUG_PUTS("Filter not found");
+            return -1;
+        }
+    }
+    else {
+        _filter_exists = _find_filter(dev, filter, true, &idx);
+        if(_filter_exists) {
+            DEBUG("Standard filter to disable at idx = %d\n", idx);
+            dev->msg_ram_conf.std_filter[idx].SIDFE_0.bit.SFEC = CANDEV_SAMD5X_FILTER_DISABLE;
+        }
+        else {
+            DEBUG_PUTS("Filter not found");
+            return -1;
+        }
+    }
+
+    return idx;
+}
+
+static int _set(candev_t *candev, canopt_t opt, void *value, size_t value_len)
+{
+    can_t *dev = container_of(candev, can_t, candev);
+    int res = 0;
+
+    switch (opt) {
+        case CANOPT_BITTIMING:
+            if (value_len < sizeof(struct can_bittiming)) {
+                return -1;
+            }
+            else {
+                memcpy(&candev->bittiming, value, sizeof(struct can_bittiming));
+                uint32_t clk_freq = sam0_gclk_freq(dev->conf->gclk_src);
+                can_device_calc_bittiming(clk_freq, &bittiming_const, &candev->bittiming);
+                res = _init(candev);
+                if (res == 0) {
+                    res = sizeof(candev->bittiming);
+                }
+                else {
+                    return -1;
+                }
+            }
+            break;
+        case CANOPT_RX_FILTERS:
+            if (value_len < sizeof(struct can_filter)) {
+                return -1;
+            }
+            else {
+                res = _set_filter(candev, value);
+                if (res >= 0) {
+                    res = sizeof(struct can_filter);
+                }
+                else {
+                    return -1;
+                }
+            }
+            break;
+        case CANOPT_STATE:
+            if (value_len < sizeof(canopt_state_t)) {
+                return -1;
+            }
+            else {
+                switch(*((canopt_state_t *)value)) {
+                    case CANOPT_STATE_OFF:
+                        res = _power_off(dev);
```

Comment:
You should control the transceiver standby pin here

Reply:
```suggestion
                        res = _power_off(dev);
                        if( gpio_is_valid(dev->conf->standby_pin) ) gpio_set((dev->conf->standby_pin)) 
                        if( gpio_is_valid(dev->conf->sleep_pin) ) gpio_set((dev->conf->sleep_pin))
                        if( gpio_is_valid(dev->conf->silent_pin) ) gpio_set((dev->conf->silent_pin))
```
just make sure you also handle FAULT and WAKE where they need to be handled
![image](https://github.com/RIOT-OS/RIOT/assets/814189/55535187-e291-4415-a856-e5414652ac07)

Reply:
I'm against adding the CAN transceiver pins to the CAN controller structure

Reply:
The driver controls/owns the transceiver, there shouldn't be any outside code fiddling with the transceiver without the driver knowing about it.

Reply:
You could argue that we could have transceiver drivers that are then attached to a CAN driver (like the PHY subsystem on Linux), but it should not be the job of the application to deal with all possible transceiver variants.

Reply:
but it is even less off a thing for the periph driver to deal with, the board designer might choose another transceiver (sleep, silent, fault or nc ) or even tie that pin to gnd (seems like the typical mcp2512 board does that)
why would the periph driver care for that nonsense

Reply:
If there is no such pin, just put `GPIO_UNDEF` there and don't handle it.

If the CAN communication is not handled by an app but by a module that is shared across multiple projects, do you really want to put board specific ifdefs in there?

The CAN API provides power states that map to the power states provided by CAN transceivers, if we set e.g. `CANOPT_STATE_SLEEP` I would expect the can transceiver to also enter standby mode without having to fiddle with a board specific pin.

Reply:
> You could argue that we could have transceiver drivers that are then attached to a CAN driver (like the PHY subsystem on Linux), but it should not be the job of the application to deal with all possible transceiver variants.

That's definitely how it should be done, because CAN controller and CAN transceiver are two different components on different communication layers. Is there a physical layer framework that can be used here and where to register the CAN transceiver driver?

Reply:
seem like we have 'can_trx' and  'drivers_tja1042' implements it  (standby only) and ncv7356 for more features,

seems like we need a driver for the `ATA6561` since that one definitely is no `tja1042` 
(over specific driver incoming TCAN1042 TCAN1043 NXP also has more parts :) )

Reply:
eh we can just use the `tja1042` driver if it's compatible.

Reply:
Can we move this conversation to #20480?

Reply:
this is resolved in favor of the separate cantrx-driver.